### PR TITLE
Fixes #497: User Preferences bugs, refactor form logic for phone settings

### LINKED ISSUE #497: User preferences Phone number bugs
Address multiple regressions between QA and production on the user preferences page, where users can save and edit their phone information & text messaging preferences. Most likely introduced in PR for #391, so find out if some or all of these bugs can be addressed:
- clearing (backspace) and saving the empty phone number shows "your preferences were not saved" in QA. In production, does this properly remove the phone number for the user?
- unchecking "I want to be notified via SMS" shows preferences not saved - investigate how checking checkboxes gets saved 
- trying out an invalid number created an empty warning message

### PR DESCRIPTION
**Pull request by jaydonkrooss**
Pretty significant rework of the user preferences to address tricky bugs. Uses a new child component and more descriptive state for organizing form logic. Now should be properly responsive to different types of changes, including phone verification, clearing settings, and notification checkboxes.

See comments in #497 for test cases that properly address changes.

### CODE REVIEW

**Comment by jxiao21** [comment]
Did this branch work off of the updated version of the master branch? I'm running into Twilio issues again when I check into this branch.

**Comment by jaydonkrooss** [comment]
@jxiao21 Forgot to pull in the changes, you can try it out now

File: src/assets/src/components/preferences.tsx [hunk 1]
```diff
@@ -12,52 +11,42 @@ import { usePromise } from "../hooks/usePromise";
...
+    const [countryDialCode, setCountryDialCode] = useState(props.user.phone_number !== "" ? props.user.phone_number.substring(0, 1) :"");
     const [notifyMeAttendee, setNotifyMeAttendee] = useState(props.user.notify_me_attendee);
     const [notifyMeHost, setNotifyMeHost] = useState(props.user.notify_me_host);
-    const [validationStatus, setValidationStatus] = useState(undefined as undefined | ValidationStatus);
+    const [formStatus, setFormStatus] = useState(FormStatus.NotSubmitted);
+    const [validationErrors, setValidationErrors] = useState([] as Error[]);
+
+    useEffect(() => {
+        if (validationErrors.length) setFormStatus(FormStatus.ValidationErrors)
+            // else setFormStatus(FormStatus.NotSubmitted)
```
Comment (jxiao21) [inline src/assets/src/components/preferences.tsx]: Remove commented out code

File: src/assets/src/components/preferences.tsx [hunk 2]
```diff
@@ -79,225 +68,103 @@ function PreferencesEditor(props: PreferencesEditorProps) {
...
-        // check if time buffer has passed, phone number has changed, 
-        // and phone number is valid
-        const timeRemaining = timeToResendCode ?
-            [new Error(`You must wait ${timeToResendCode} more seconds before requesting a new verification code.`)]
+    const validateAndSubmit = (e: React.SyntheticEvent) => {
+        e.preventDefault() // Prevent page reload
+        // if (!changedPhoneNumber && !changedNotifications) {
+        //     setFormStatus(FormStatus.NoChanges);
+        //     return;
+        // }
```
Comment (jxiao21) [inline src/assets/src/components/preferences.tsx]: Commented out code should be removed

File: src/assets/src/components/preferences.tsx [hunk 3]
```diff
@@ -79,225 +68,103 @@ function PreferencesEditor(props: PreferencesEditorProps) {
...
-        e.preventDefault() // Prevent page reload
-        if (!changedNotificationSettings) {
-            setValidationStatus(null);
-            return;
+    const alertBlock = (() => {
+        switch (formStatus) {
+            case FormStatus.NotSubmitted:
+                return undefined;
+            // case FormStatus.NoChanges:
+            //     return <Alert variant='primary'>No preferences were changed.</Alert>;
```
Comment (jxiao21) [inline src/assets/src/components/preferences.tsx]: commented out code should be removed

**Comment by jxiao21** [comment]
Ah I see you just went through and removed the unused code while I was reviewing it haha

**Review by jxiao21 — APPROVED** [review]
lgtm

**Pull request merged**